### PR TITLE
Refactor branch walk #1175

### PR DESCRIPTION
**Pull request by theoreticalbts**
PR for #1175

### CODE REVIEW

File: libraries/chain/database.cpp [hunk 1]
```diff
@@ -500,6 +500,23 @@ bool database::push_block(const signed_block& new_block, uint32_t skip)
...
+   auto blocks = _fork_db.fetch_block_by_number( height );
+   if( blocks.size() > 1 )
+   {
+      vector< std::pair< account_name_type, fc::time_point_sec > > witness_time_pairs;
+      for( const auto& b : blocks )
+      {
+         witness_time_pairs.push_back( std::make_pair( b->data.witness, b->data.timestamp ) );
+      }
+
+      ilog( "Encountered a block num collision due to a fork. Walking the current fork to determine the correct block. block_num:${n}", ("n", height) );
```
Comment (mvandeberg) [inline libraries/chain/database.cpp]: We should combine these messages and make the logging accurate. Something like, "Encountered block num collision at block ${n} produced by witnesses ${w}" This also removes the statement that we are walking the fork, which we are not doing at this point in the code.

**Comment by theoreticalbts** [comment]
Still not building.  Will fix the build and address the above comment momentarily.

File: libraries/chain/database.cpp [hunk 2]
```diff
@@ -500,6 +500,22 @@ bool database::push_block(const signed_block& new_block, uint32_t skip)
    return result;
 }
 
+void database::_maybe_warn_multiple_production( const database& db, uint32_t height )const
```
Comment (mvandeberg) [inline libraries/chain/database.cpp]: If this method is a member of database, why pass a const reference to itself?

Comment (theoreticalbts) [inline libraries/chain/database.cpp]: Good catch, originally it wasn't, but then I saw `_fork_db` was private.

**Pull request merged**